### DISCUSSION TITLE
Parent Beta Phase Reconstruction in Titanium Alloys

### DISCUSSION THREAD

**Discussion by louis96108** [mtex-toolbox/mtex#838] [Ask Anything]
I tried to reconstruct the Beta Phase with the codes of 'Documentation',but  there is no Beta phase information in my EBSD data,so there is something wrong when I use 'job = parentGrainReconstructor(ebsd, grains, beta2alpha)'.I want to ask that how to reconstruct the parent Beta phase if there is no Beta phase information in the EBSD data. Thanks a lot.

**Comment by ralfHielscher** [comment]
Hi Louis,

in this case you have to specify the crystalSymmetry of the beta phase by hand, i.e., 
```matlab
cs_beta = crystalSymmetry('432',[3.3, 3.3, 3.3],'mineral','Ti-Beta')
```
and then you use this while defining the Burges relationship `beta2alpha`

Ralf.

**Reply by louis96108** [reply]
Hi ralfHielscher,
    Thanks for your last answer.Here is the complete code after importing the EBSD data

csBeta = crystalSymmetry('432',[3.3 3.3 3.3],'mineral','Titanium (beta)');
beta2alpha = orientation.Burger(csBeta,ebsd('Titanium (alpha)').CS);
round2Miller(beta2alpha)
[grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'threshold',1.5*degree,...
  'removeQuadruplePoints');
job = parentGrainReconstructor(ebsd, grains, beta2alpha)
job.calcTPVotes('numFit',2)
job.calcParentFromVote('strict', 'minFit',2.5*degree,'maxFit',5*degree,'minVotes',2)
job.calcGBVotes('noC2C');
job.calcParentFromVote('minFit',5*degree)
job.mergeSimilar('threshold',5*degree)
job.mergeInclusions('maxSize',50)
color = ipfKey.orientation2color(job.parentGrains.meanOrientation);
plot(job.parentGrains, color, 'figSize', 'large')


     After running these ,Matlab shows that I misuse the function' parentGrainReconstructor/calcParentFromVote' (line 30)
and points out 'You need to compute votes first.’   I don't know what's wrong with the 'job.calcTPVotes('numFit',2)'. Looking forward to your answer!Thank you.

**Reply by ralfHielscher** [reply]
After executing `job.calcTPVotes('numFit',2)` MTEX gives you some output to the command window. It would be nice if you could post it,

Ralf.

**Reply by louis96108** [reply]
>> job = parentGrainReconstructor(ebsd, grains, beta2alpha)
job.calcTPVotes('numFit',2)
job.calcParentFromVote('strict', 'minFit',2.5*degree,'maxFit',5*degree,'minVotes',2)
job.calcGBVotes('noC2C');
job.calcParentFromVote('minFit',5*degree)
 
job = parentGrainReconstructor (show methods, plot)
 
 phase   mineral           symmetry  grains  area  reconstructed
 parent  Titanium (beta)   "432"     0       0%    0%           
 child   Titanium (Alpha)  "6/mmm"   145     100%               
 
 parent to child OR: (185.3°,90°,45°)
   fit with child to child misorientations: 12°, 19°, 24°, 26° (quintiles)
   closest ideal OR: (110) || (0001)   [1-11] || [-2110]   error: 0°
 
progress: 100%
 
ans = parentGrainReconstructor (show methods, plot)
 
 phase   mineral           symmetry  grains  area  reconstructed
 parent  Titanium (beta)   "432"     0       0%    0%           
 child   Titanium (Alpha)  "6/mmm"   145     100%               
 
 parent to child OR: (185.3°,90°,45°)
   fit with child to child misorientations: 12°, 19°, 24°, 26° (quintiles)
   closest ideal OR: (110) || (0001)   [1-11] || [-2110]   error: 0°
 
 votes: 630 
   best fit: 32°, 40°, 45°, 51°         (quintiles)
   second best fit: 32°, 40°, 45°, 51°  
  votes   fit < misFit: 0.48%
  votes  fit2 > maxFit: 0%
  numVotes >= minVotes: NaN%
 
 
ans = parentGrainReconstructor (show methods, plot)
 
 phase   mineral           symmetry  grains  area  reconstructed
 parent  Titanium (beta)   "432"     0       0%    0%           
 child   Titanium (Alpha)  "6/mmm"   145     100%               
 
 parent to child OR: (185.3°,90°,45°)
   fit with child to child misorientations: 12°, 19°, 24°, 26° (quintiles)
   closest ideal OR: (110) || (0001)   [1-11] || [-2110]   error: 0°
 
The above is the output, then Matlab shows that I misuse the function' parentGrainReconstructor/calcParentFromVote' (line 30)and points out 'You need to compute votes first.’

**Reply by louis96108** [reply]
Hi ralfHielscher，
    I  think 'job.calcGBVotes('noC2C')'  is wrong . After executing ,I've got no output, just 'job.calcGBVotes('noC2C')' is shown in the command window. This may cause the misusing of the next function, but I still don't know how to correct this error.Could you tell me the methods?   
   Moreover，when I use 'color=ipfKey.orientation2color(job.parentGrains.meanOrientation)',Matlab shows that undefine variable 'ipKey'. I use Matlab2019a and MTEX 5.6. Look forward to your reply！Thank you very much!


Louis

**Reply by ralfHielscher** [reply]
Hi Louis,

you should go a bit slower through your script and make sure you understand each step.
After you initialized the  parentGrainReconstructor with 

```matlab
job = parentGrainReconstructor(ebsd, grains, beta2alpha)
```
you get
```
job = parentGrainReconstructor (show methods, plot)

phase mineral symmetry grains area reconstructed
parent Titanium (beta) "432" 0 0% 0%
child Titanium (Alpha) "6/mmm" 145 100%

parent to child OR: (185.3°,90°,45°)
fit with child to child misorientations: 12°, 19°, 24°, 26° (quintiles)
closest ideal OR: (110) || (0001) [1-11] || [-2110] error: 0°
```
This should already alert you, as it says that in your data set are almost no alpha 2 alpha grain boundaries.
Possible reasons:
* wrong OR
* no childs in the dataset
* wrongly  imported data.
You need to investigate this. Pick a boundary where you think it is child 2 child and compare the misorientation with the theorerical one.

Ralf.

**Reply by Savillian** [reply]
Hi all,

I've also run into the same issue, and it seems to be a function of the import file not having Beta-Ti listed as a possible phase. I've tried the reconstruction process with both MTEX 5.5 and 5.6 processes, but to no avail on both.

As a workaround until I figure out exactly why the data isn't being voted correctly, I plan on running all my EBSD collection with Alpha and Beta Ti even if I don't have the latter phase present.

@ralfHielscher - I was reading up on the martensite --> austenite reconstruction process, and I noted it operates in a different fashion to the titanium reconstruction process. Could this approach be more applicable for single phase reconstructions than this path?

Alec

**Reply by ralfHielscher** [reply]
Hi all,

would it be possible for you to send me the data? it is very hard to tell from distance whats going. From my point of view having only the alpha phase available should not cause any problems.

This approach can be applied as well, sure. Whether it is "more" applicable is hard to say.

Ralf.

**Reply by louis96108** [reply]
Hi all ,
   A few days ago Mr.ralfHielscher said maybe there is something wrong when I import the EBSD Data,so I've pasted the output of the ’import EBSD Data' as follow ,the output of reconstruct grains has also be pasted . I checked the figure of grain boundary   and I think the alpha phase boundary information has been imported into the MTEX. If this step is no problem ,I will send my data to Mr.ralfHielscher.

%% Import the Data

% create an EBSD variable containing the data
ebsd = EBSD.load(fname,CS,'interface','ctf',...
  'convertEuler2SpatialReferenceFrame')
 
ebsd = EBSD (show methods, plot)
 
 Phase  Orientations           Mineral         Color  Symmetry  Crystal reference frame
     0    20 (0.21%)        notIndexed                                                 
     1   9732 (100%)  Titanium (Alpha)  LightSkyBlue     6/mmm       X||a*, Y||b, Z||c*
 
 Properties: bands, bc, bs, error, mad, x, y
 Scan unit : um
 
>> [grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'threshold',20*degree,...
  'removeQuadruplePoints')
 
grains = grain2d (show methods, plot)
 
 Phase  Grains  Pixels           Mineral  Symmetry  Crystal reference frame
     1      78    9732  Titanium (Alpha)     6/mmm       X||a*, Y||b, Z||c*
 
 boundary segments: 1878
 inner boundary segments: 0
 triple points: 125
 
 Properties: GOS, meanRotation
 
 
ebsd = EBSD (show methods, plot)
 
 Phase  Orientations           Mineral         Color  Symmetry  Crystal reference frame
     0    20 (0.21%)        notIndexed                                                 
     1   9732 (100%)  Titanium (Alpha)  LightSkyBlue     6/mmm       X||a*, Y||b, Z||c*
 
 Properties: bands, bc, bs, error, mad, x, y, grainId
 Scan unit : um


Louis

**Reply by ralfHielscher** [reply]
From this no error or mistake can be seen.

Ralf.

**Reply by Savillian** [reply]
Hi all,

I've isolated the issue to something with the initial voting process. I can correctly generate the IPF orientation maps of my data, and checking the "ebsd" variable returns everything correctly. However, the first voting command returns no votes. I did get the process to work partially once (I still don't know what changed from the previous attempt), but the reconstruction was incomplete and still requires more troubleshooting. It has since stopped working again...

@louis96108 - Are you doing any kind of filtering/smoothing of your data before processing?

On a separate note, another issue emerged when I created a new script to troubleshoot the processing. My beta2alpha variable returns the following error despite my IPF orientation maps being loaded correctly just before.

-------------------------------------------------------------------------
Error using phaseList/checkSinglePhase (line 317)
There are no indexed data in this variable!

Error in phaseList/get.CS (line 154)
      id = checkSinglePhase(pL);

Error in EBSD/subsref (line 40)
  [varargout{1:nargout}] = builtin('subsref',ebsd,s);

Error in MTEXBetaReconstructionUD560 (line 82)
beta2alpha = orientation.Burger(CSB,ebsd('Titanium (Alpha)').CS);;


_Where CSB = crystalSymmetry('m-3m', [3.3 3.3 3.3], 'mineral', 'Ti (Beta)', 'color', [0.56 0.74 0.56]);_

@ralfHielscher I'll email you my data file momentarily. Thanks again for all your help!

Alec

**Reply by louis96108** [reply]
Hi Alec
   I just run these fuctions   
   [grains,ebsd.grainId] = calcGrains(ebsd('indexed'),'threshold',20*degree,...
  'removeQuadruplePoints')
grains = smooth(grains,5);


Louis

**Reply by Savillian** [reply]
Hi Louis,

Thanks for clarifying, I thought this might have been where the issue was arising but that doesn't seem to be the case now...

Have you had any luck since posting?

All the best,
Alec

**Reply by Savillian** [reply]
Hi @louis96108 and @ralfHielscher,

I may have found the issue. I was employing a pre-processing script akin to that @louis96108 included here, and this seems to be causing the reconstruction process to fail. Something inside the smoothing script is changing the ebsd variable structure, and returning a series of values useless for the reconstruction process. Removing this operation enables my reconstruction to complete. The workaround here I think is to have a separate matlab script for generating alpha-Ti plots and then a second script just for the reconstruction without pre-processing. Smoothing could be applied after the reconstruction process I expect, but I haven't tried this yet.

I haven't isolated why this was the case though. I've included the script I removed below for reference. Do either of you know why this would cause the issue? 

[grains,ebsd.grainId,ebsd.mis2mean] = calcGrains(ebsd,'angle',2*degree);
ebsd(grains(grains.grainSize<10)) = [];
[grains,ebsd.grainId] = calcGrains(ebsd,'angle',1.5*degree);
F = halfQuadraticFilter;
F.alpha = 0.5;
ebsd2 = smooth(ebsd,F,'fill',grains);
[grains] = calcGrains(ebsd2,'angle',2*degree);

Alec

**Reply by Savillian** [reply]
Hi all,

Another brief update, I think I found the issue. I was using too large of a value for removing small grains and this was destroying the integrity of any Alpha-Ti grains. Thus, when I tried to vote on the orientation reconstruction, there was nothing for the script to operate off of.

All the best,
Alec
